pkg/providers/vsphere/network/network.go

```diff
@@ -213,9 +209,34 @@ func applyInterfaceSpecToResult(
 		result.DHCP6 = true
 	}
 
-	// For VPC we set the interface spec IPs in the SubnetPort so ignore those addresses
-	// here. Otherwise, even for NoIPAM still honor the interface spec addresses.
-	if !isVPC && len(interfaceSpec.Addresses) > 0 {
+	if len(interfaceSpec.Addresses) > 0 {
+		if interfaceSpec.Gateway4 == "" || interfaceSpec.Gateway6 == "" {
+			// Backfill the gateways from the network provider if not specified in
+			// the interface spec before setting the user specified addresses. This
+			// allows for just IP reservation to be done.
+			// Our result is modeled after the network interface CRs, but that ends
+			// up being cumbersome here. Instead, the gateways should be pulled up
+			// in the top level results instead of being a part of the IP config.
+			for i := range result.IPConfigs {
+				gw := result.IPConfigs[i].Gateway
+				if gw == "" {
+					continue
+				}
+
+				if interfaceSpec.Gateway4 == "" { //nolint:gocritic
```

Comment:
would something like this help get rid of the //nolint:gocritic?

    for _, ipConfig := range result.IPConfigs {
        if ipConfig.Gateway == "" {
            continue
        }
        
        if ipConfig.IsIPv4 && interfaceSpec.Gateway4 == "" {
            interfaceSpec.Gateway4 = ipConfig.Gateway
        }
        if !ipConfig.IsIPv4 && interfaceSpec.Gateway6 == "" {
            interfaceSpec.Gateway6 = ipConfig.Gateway
        }
        
        if interfaceSpec.Gateway4 != "" && interfaceSpec.Gateway6 != "" {
            break
        }
    }